coffee_core/src/coffee.rs

```diff
@@ -272,11 +273,77 @@ impl PluginManager for CoffeeManager {
         })
     }
 
-    async fn upgrade(&mut self, _: &[&str]) -> Result<(), CoffeeError> {
+    async fn upgrade_single(&mut self, repo: &str) -> Result<CoffeeUpgradeOne, CoffeeError> {
+        let status;
+        match self.repos.get(repo) {
+            Some(repo) => {
+                // get the list of the plugins installed from this repository
+                // TODO: add a field of installed plugins in the repository struct instead
+                let mut installed_plugins: Vec<String> = vec![];
+                let remote_repo = repo.list().await?;
+                let plugins = self.config.plugins.clone();
+                for plugin in &remote_repo {
+                    if let Some(ind) = plugins
+                        .iter()
+                        .position(|elem| elem.name() == *plugin.name())
+                    {
+                        let plugin_name = &plugins[ind].name().clone();
+                        installed_plugins.push(plugin_name.to_owned());
+                    }
+                }
+                // pull the changes from the repository
+                status = repo.pull().await?;
+                match status {
+                    UpgradeStatus::UpToDate => {
+                        log::debug!("repository up to date: {}", repo.name());
+                    }
+                    UpgradeStatus::Updated => {
+                        // uninstall the plugins
+                        // for plugin in &installed_plugins {
+                        //     self.remove(&plugin).await?;
+                        // }
+                        // reinstall the plugins
+                        // for plugin in &installed_plugins {
+                        //     self.install(&plugin, false, false).await?;
+                        // }
```

Comment:
But then if a plugin changed the name (path) of the executable in an upgrade/branch, we would have a problem because the older version of the exec_path would still exist in the configuration.